project/__init__.py

```diff
@@ -98,6 +108,63 @@ def push(self):
             self.builddir)
         ]
 
+    def pull_request(self):
+        """
+        Creates a pull request from the "update" branch to master.
+        """
+        authorization_header = {"Authorization": "token " + self.github_token}
+
+        pulls_api_url = 'https://api.github.com/repos/platformsh-templates/{0}/pulls'.format(self.name)
+
+        body = {"head": "update", "base": "master", "title": "Update to latest upstream"}
+        response = requests.post(pulls_api_url, headers=authorization_header, data=json.dumps(body))
+        return response.status_code in [201, 422]
+
+    def test(self):
+        """
+        Calls all the tests defined in self.TEST_FUNCTIONS on the test urls.
+        """
+        urls_to_test = self.get_test_urls()
+        if not urls_to_test:
+            print("No pull requests to test for {0}".format(self.name))
+
+        results = []
+        print(f"Running {len(self.TEST_FUNCTIONS)} tests.")
+        for test in self.TEST_FUNCTIONS:
+            for url in urls_to_test:
+                results.append(test(url))
+
+        return all(results)
+
+    def merge_pull_request(self):
```

Comment:
I don't think we want any automatic merging of PRs.  PRs should always have human validation.  Create the PR, sure, but humans should be the ones doing all merges (since there is often testing needed that goes beyond what can be automated easily).

Reply:
It could be left in the code as a feature to be triggered manually instead of by the full pipeline. Or do we want that feature completely removed?

Reply:
The whole point of this exercise is to fully automate the template updating process (running in travis CI daily) and take everything out of human hands. You can't have human interaction in CI. If there are tests that are non-trivial to automate there will be non-trivial implementations to be made. "Every pull request should have human validation" is a valid point regarding code changes that need _review_, but here it's a matter of updating dependencies and no human needs to sift through those. On the engineering side, all vendored packages and container images get automated releases and I don't see how this is any different.

Reply:
I'm completely on board with automation replacing 90% of the process.  But I don't have faith that automated tests will be able to fully verify that a given change doesn't break.  Eg, right now I am, once again, tracking down an intermittent issue with Drupal's installer completing successfully but not resulting in a working site at the end, sometimes.  I don't expect automation to be able to step through an interactive Ajax based installer and then verify that the resulting site works correctly.

Even just updating dependencies can cause something to break.  I would much prefer the merge button always be pressed by a human, even if everything before that is a nightly automated task.

The dependencies underpinning some of these applications are far, far less stable than the ones that underpin the container images.  I wish that weren't true, but...